chore: add my blogpost (#617)
Effect-TS/website: content/blog/0021-bundle-accounting.mdx

- `Scope`
- `ManagedRuntime`
- `GlobalValue`
- `String`
- `HttpClient`
- `ClientRequest`
- `ClientResponse`
- `Schema`
- `Data`

"Oh no! Egad! Thats too much to learn!" You're saying right now, totally ignoring that you probably have learned all of these concepts, and much more, piecemeal, from scratch on every project according to whatever 30-or-so packages are included in the project.

Don't worry dear reader, you don't have to learn any of these, you can be productive
with `effect` even when using just a single module, and learn to use the rest at your own pace. For me its the missing standard library that is internally consistent and 100% interoperable.

Cut to yesterday, I've finally finished implementing all the features I needed. 80 changed files (mostly new files), around 12K LOC added by yours truly (trimmed a lot of fat thanks to `effect`, but react theres still a lot of React code and CSS).

So, nows the money time. How much did I pay for my hubris?

[net 50kb difference](/public/images/datner-bundle-50kb.png)

Net 50kb.

`effect`, `@effect/schema`, `@effect/platform`, `@effect/platform-browser`, and `react-query` to boot.

80 files, 4 new routes, nearly all new code.

> 

Reviewer: ```suggestion

As `effect` makes its way into more and more areas of your application, its size amortizes because you inevitably trim a lot of "fat" from your application. I removed almost all the custom utilities I had previously built and was able to uninstall several packages.

Additionally, `effect` is 100% tree-shakeable, so I didn't end up with unnecessary code in my final application bundle.

I can say without hesitation that the `effect` "tax" was 100% worth it.
```

Author: a bit too plastic-y, don't you think?

Reviewer: Edit as you see fit :) this was just a suggestion.

Don't worry dear reader, you don't have to learn any of these, you can be productive
with `effect` even when using just a single module, and learn to use the rest at your own pace. For me its the missing standard library that is internally consistent and 100% interoperable.

Cut to yesterday, I've finally finished implementing all the features I needed. 80 changed files (mostly new files), around 12K LOC added by yours truly (trimmed a lot of fat thanks to `effect`, but react theres still a lot of React code and CSS).

So, nows the money time. How much did I pay for my hubris?

[net 50kb difference](/public/images/datner-bundle-50kb.png)

Net 50kb.

`effect`, `@effect/schema`, `@effect/platform`, `@effect/platform-browser`, and `react-query` to boot.

80 files, 4 new routes, nearly all new code.

This was not an isolated case, as `effect` makes its way into more and more areas of your application, its size amortizes because you inevitably trim a lot of "fat" from your application. I removed almost all the custom utilities I had previously built and was able to uninstall several packages.

> Grief and effort was saved from implementing and fixing bugs in tools that I get for free just by using `effect`

Reviewer: ```suggestion

Grief and effort was saved from implementing and fixing bugs in tools that I get for free just by using `effect`.
```

> Additionally, `effect` is 100% tree-shakeable, so I didn't end up with unnecessary code in my final application bundle.

Reviewer: ```suggestion

Additionally, `effect` is 100% tree-shakeable, so I didn't end up with unnecessary code in my final application bundle.
```

# The Effect Tax

It's been over a year now since I took the effect pill and I'll probably never develop the same way as I did before. I love effect, and I am absolutely a power-user.

But I am in my core a frontend oriented developer. I care about bundle size and perceived performance, and while the latter was never a concern to me as I've seen in my own eyes very complex application running at 120FPS powered by effect and react, the former was just a fact of matter-- effect is *large*

A couple of weeks ago I inherited a `create-react-app` project, yes, shudder.
It had all the classics, AnyScript, class components, esoteric 1-off deps preventing updates,
a million lint violations of the `==` kind, multiple overlapping component libraries,
and multiple style solutions.

> My task was to add 4 pages and a whole bunch of features.
> Using the existing material was out of the question, I **had** to create my own lib from scratch
> if I wanted to have any semblence of confidence in the app.

Reviewer: ```suggestion
My task was to add four pages to the app as well as a whole bunch of features. Using the existing material was absolutely out of the question - I would have **had** to create my own lib from scratch if I wanted to have any shred of confidence in the app.
```

> I started hacking at it. It didn't take a full hour before I had to wrangle untyped garbage from the legacy
> side of the fence. It was an http call, it wasn't parsed, there was no `zod` or something. It wasn't
> even typed `Promise<Something>`, it was proudly `Promise<any>` and then just consumed like thing.
> I wasn't there at the dawn of creation, I had no idea what the previous dev thought the shape was
> I could only follow function names and see the properties being arbitrarily accessed. So I did research to
> find out what the shape was supposed to be and created a util:

Reviewer: ```suggestion
I started hacking, and it didn't even take a full hour before I had to wrangle untyped garbage from the legacy side of the fence for a simple HTTP call. Here were some of the problems I found almost immediately: 
- It wasn't parsed (i.e. there was no `zod` or other schema solution)
- It wasn't even typed `Promise<Something>`, it was proudly typed as `Promise<any>` and then just consumed the untyped result
- I wasn't there at the dawn of creation of the app, so I had no idea what the shape of the result even was - I could only follow function names and see the properties being arbitrarily accessed.

So I did research to find out what the shape was supposed to be and created a little utility:
```

```ts
const UnsafeTypeId: unique symbol = Symbol.for("@types/Unsafe");

/**
 * Brands `A` as an unsafe type.
 * An unsafe type is a type that was cast but never validated.
 *
 * @example
 * ```ts
 * const response = await axios.get(...)
 *
 * const data = response.data as Unsafe<Todo>
 * ```
 *
 * @author Datner<contact@datner.me>
 */
export type Unsafe<A> = A & {
  readonly [UnsafeTypeId]?: "Unsafe";
};
```

> Now I can give a proper shape and acknowledge that it is infact a bullshit promise, I never made sure that
> The shape is actually that, but it's more than `any`.
> Then I encountered another situation that prompted a different util, and then another, and then
> even more, I found myself building all the
> inftrastructure and utils for the basic essentials of just making usable software.
> I tried coping, letting go, but it just kept coming back and it dawned on me. I'm just reimplemeting effect.

Reviewer: ```suggestion
Now I can at least give a proper shape to the result while also acknowledging to future developers that the shape of the type wrapped by `Unsafe` is not actually validated. 

Then I encountered another situation that prompted me to implement a different utility ... and then another ... and then *another* ... and on and on it went. I found myself implementing infrastructure and utilities just so I would have the basic essentials to write usable software.

First I tried just coping with it, then I tried to just let go and get the project done, but these issues continued to linger in the back of my mind ... and then it dawned on me - I was just re-implementing `effect`.
```

> Worse, I'm reimplemeting effect by trying to hot-glue libs together. Oh the axios client needs the current
> tenant so it needs access to the context but it can change in the lifetime of the application and it also
> needs the bearer token but it might need revalidation or even a relog oh but I make too many requests
> for the same thing I need some caching--- It's been nearly a month
> now and I've only been building inftrastructure!
> The boring kind too! Plus I hate the result and it's full of bugs I won't see until they pop up in production!

Reviewer: ```suggestion
Worse, I'm re-implementing `effect` by trying to glue libraries together that were never designed to work with one another. 

Oh the axios client needs the current tenant? Ok, then it needs access to the context. But it can change in the lifetime of the application and also needs the bearer token, but it might need revalidation or even a re-log. Ah, but I make too many requests for the same thing so I also need some caching. It's been nearly a month now and I've only been building infrastructure! The boring kind too! Plus I hate the result and it's full of bugs I won't see until they pop up in production!
```

> I gave up. Whatever the effect tax is, it's worth it.
> From what we've seen, the effect tax is about 50kb. Depending on what you feel this can be
> a lot or very little, but considering that I added around 20k LOC to the project, plus new dependencies
> It was not worth considering.

Reviewer: ```suggestion
So I gave up. Whatever the `effect` "tax" is, it's worth it. From what we've seen, the "tax" on bundle size is about 50kb. Depending on your use-case, this may seem like a lot or very little. But considering that I added around 20k LOC to the project, as well as some new dependencies, for me it wasn't even a consideration.
```

```sh
$ npm i effect @effect/schema @effect/platform @effect/platform-browser
```

> (yes this project used npm, no I don't recommend)

Reviewer: ```suggestion
```

> I've started replacing all my imitations with the utils from effect.
> To scare ya a bit, heres a `combine` function I used for a particular `useQueries`
> from the great `@tanstack/react-query` (which I also added a week ago)

Reviewer: ```suggestion
I began replacing all my imitations with the utils from `effect`. 

Here is a `combine` function I created for a particular `useQueries` query from the great `@tanstack/react-query` (a dependency which I also added a week ago).
```

```ts
import { Array, Predicate, Option, pipe } from "effect"

// further down....
const combine = (
  queries: [
    UseSuspenseQueryResult<User[], Error>,
    UseSuspenseQueryResult<Recipient[], Error>,
  ],
) => {
  const [users, recipients] = queries;

  const initialRows = pipe(
    users.data,
    Array.filter((_) => Option.isSome(_.emailAddress)),
    Array.appendAll(recipients.data),
    Schema.decodeSync(Schema.Array(NormalizedRow)),
    Array.dedupeWith((a, b) => a.Email === b.Email && a.fromISP),
  );

  const isRecipient = Predicate.or(RecipientRow.isRecipient, (_) =>
    recipients.data.some((r) => r.recepientId === _.id),
  );
  return {
    pending: queries.some((_) => _.isPending),
    initialRows,
    initialSelected: Array.filterMap(initialRows, (_) =>
      isRecipient(_) ? Option.some(_.id) : Option.none(),
    ),
  };
};
```

> This function takes the result of two collections that have some possible overlap,
> cleans up `users` from members that don't have an email,
> concatenates it to the `recipients` into a nice wholesome `(User | Recipient)[]` thats unusable,
> uses an effect schema `NormalizedRow` to homogenize it into a useful `RecipientRow[]`, and remove duplicate
> rows, keeping only the `User`-originated ones.

Reviewer: ```suggestion
This function takes the result of two collections that have some possible overlap, cleans up `users` from members that don't have an email, concatenates it to the `recipients` into a nice wholesome `(User | Recipient)[]` that is unusable, uses an `@effect/schema` `NormalizedRow` schema to homogenize it into a useful `RecipientRow[]`, and remove duplicate rows, keeping only the `User`-originated ones.
```

> Then also make a collection derived from it thats just the ids of the `RecipientRows` that either represent
> a `Recipient` *or* representa `User` *that has a corresponding `Recipient`* (remember we deduped them)

Reviewer: ```suggestion
This utility also then creates a derived collection that is just the ids of the `RecipientRows` that either represent a `Recipient` *or* represent a `User` *that has a corresponding `Recipient`* (remember we deduped them).
```

> You could glue a solution that does this using `lodash`, `zod`, and some hand-crafted stuff, but it won't
> be this elegant considering the definition above. Especially the normalization part and the types.
> In this snippet it looks like it's trivial-- It's not. This function was over 100LOC before.

Reviewer: ```suggestion
You could glue a solution that does this using `lodash`, `zod`, and some hand-crafted stuff, but it won't be this elegant considering the definition above. Especially the normalization part and the types. In this snippet it looks like it's trivial -- it's not. This function was over 100 LOC before.
```

> In social media I see effect being bashed called "a different language" or "hard to learn" or "unreadable"
> mainly by people who never used the lib or have used it in bad faith to affirm their preconcived notions.
> Effect is a full toolkit for develpoing enterprise-grade applications. Not an RxJS simulator or a cool way
> to use `Result` from rust.

Reviewer: ```suggestion
Recently, I have seen `effect` being referred to as "a different language" or "hard to learn" or "unreadable", including commentary from people who have never even used the library. Effect is a complete toolkit for developing enterprise-grade applications. Not an RxJS simulator or a cool way to use `Result` from Rust.
```

> I also have a ton of `Effect`-effect usage.
> With all the bells and whistles! From a quick search heres a partial list of modules I use in this project

Reviewer: ```suggestion
I also have a ton of `Effect`-effect usage. With all the bells and whistles! From a quick search here's a partial list of modules I use in this project
```

- `Effect`
- `Layer`
- `Context`
- `Predicate`
- `Array`
- `Cause`
- `Option`
- `Config`
- `Scope`
- `ManagedRuntime`
- `GlobalValue`
- `String`
- `HttpClient`
- `ClientRequest`
- `ClientResponse`
- `Schema`
- `Data`

> "Oh no! Egad! Thats too much to learn!" You're saying right now, totally ignoring that you learned
> all of this, and much more, piece-meal from scratch on every project according to whatever 30-or-so packages
> are included in the project.

Reviewer: ```suggestion
"Oh no! Egad! Thats too much to learn!" You're saying right now, totally ignoring that you probably have learned all of these concepts, and much more, piecemeal, from scratch on every project according to whatever 30-or-so packages are included in the project.
```

> Don't worry dear reader, you don't have to learn any of these, you can be dangerous
> even with just a single module and use the rest at your own pace. For me its the missing standard library
> that is internally consistent and interopable.

Reviewer: ```suggestion
Don't worry dear reader, you don't have to learn any of these, you can be productive
with `effect` even when using just a single module, and learn to use the rest at your own pace. For me its the missing standard library that is internally consistent and 100% interoperable.
```

> Cut to yesterday, I've finally finished implementing all the features I needed.
> 80 changed files (mostly new files), around 12K lines of code added by yours truly
> (trimmed a lot of fat thanks to effect, but react theres still a lot of react code + css).
> So, nows the money time. How much did I pay for my hubris?

Reviewer: ```suggestion
Cut to yesterday, I've finally finished implementing all the features I needed. 80 changed files (mostly new files), around 12K LOC added by yours truly (trimmed a lot of fat thanks to `effect`, but react theres still a lot of React code and CSS).

So, nows the money time. How much did I pay for my hubris?
```

[net 50kb difference](/public/images/datner-bundle-50kb.png)

> net 50kb.
> effect, schema, platform + platform-browser, and react-query to boot.
> 80 files, 4 new routes, nearly all new code.

Reviewer: ```suggestion
Net 50kb.

`effect`, `@effect/schema`, `@effect/platform`, `@effect/platform-browser`, and `react-query` to boot.

80 files, 4 new routes, nearly all new code.
```
